docs/development/design/fsm.md

```diff
@@ -0,0 +1,57 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> IDLE
+    IDLE --> AUTH: "login"
+    AUTH --> READY: "login_success"
+    AUTH --> IDLE: "login_failed"
+    READY --> PEERING: "start"
+    READY --> IDLE: "logout"
+    PEERING --> READY: "stop"
+    PEERING --> READY: "peering_error"
+    PEERING --> IDLE: "authentication_error"
+    PEERING --> IDLE: "logout"
+```
+
+nexd may be in one of four states:
+
+1. IDLE: No configuration has been provided
+2. AUTH: Performing authentication, either interactively or via token
+3. READY: Authenticated and ready to peer
+4. PEERING: Actively Peering
+
```

Comment:
Not sure if these states resonate better or not:

1. DISCONNECTED: No configuration has been provided, not logged in
2. CONNECTING: Performing authentication, either interactively or via token
3. DOWN: Authenticated and ready to bring the wg interface up
4. UP: wg interface is up and is peering.

Reply:
I like `DISCONNECTED` and `CONNECTING.

I like `READY` and `PEERING` more than `DOWN` and `UP`.

It seems like there should be one other state after `PEERING` is complete? Once all peer changes have been applied, I think it would be nice to go to a steady state like `UP`. If it receives updates that need to be applied, then it could go back into `PEERING` while it works on applying those changes.